src/models/kernels.hh

```diff
@@ -0,0 +1,250 @@
+    return std::static_pointer_cast<internal::KernelImpl<Type>>(kernel_impl);
+  }
+
+  class Kernel {
+   public:
+    using Hypers_t = typename internal::KernelImplBase::Hypers_t;
+
+    explicit Kernel(const Hypers_t & hypers) {
+      using internal::KernelType;
+      using internal::TargetType;
+
+      this->parameters = hypers;
+      if (hypers.count("representation_identifiers") == 1) {
+        for (auto & id : hypers["representation_identifiers"]) {
+          identifiers.emplace_back(id.get<std::string>());
+        }
+      }
+
+      if (hypers.count("target_type") == 1) {
+        if (hypers["target_type"] == "Structure") {
+          this->target_type = TargetType::Structure;
+        } else if (hypers["target_type"] == "Atom") {
+          this->target_type = TargetType::Atom;
+        }
+      } else {
+        throw std::runtime_error(R"(target_type is either structure or atom)");
+      }
+
+      auto kernel_type_str = hypers.at("name").get<std::string>();
+      if (kernel_type_str.compare("Cosine") == 0) {
+        this->kernel_type = KernelType::Cosine;
+        this->kernel_impl = make_kernel_impl<KernelType::Cosine>(hypers);
+      } else {
+        throw std::logic_error("Requested Kernel \'" + kernel_type_str +
+                               "\' has not been implemented.  Must be one of" +
+                               ": \'Cosine\'.");
+      }
+    }
+
+    template <class Calculator, class StructureManagers>
+    math::Matrix_t compute(const Calculator & calculator,
+                           const StructureManagers & managersA,
+                           const StructureManagers & managersB) {
+      using ManagerPtr_t = typename StructureManagers::value_type;
+      using Manager_t = typename ManagerPtr_t::element_type;
+      using Property_t = typename Calculator::template Property_t<Manager_t>;
+      auto && representation_name{calculator.get_name()};
+      using internal::TargetType;
+
+      switch (this->target_type) {
+      case TargetType::Structure: {
+        return this->compute_helper<Property_t, TargetType::Structure>(
+            representation_name, managersA, managersB);
+        break;
+      }
+      case TargetType::Atom: {
+        return this->compute_helper<Property_t, TargetType::Atom>(
+            representation_name, managersA, managersB);
+        break;
+      }
+      default:
+        throw std::logic_error("The combination of parameter is not handdled.");
+        break;
+      }
+    }
+
+    template <class Property_t, internal::TargetType Type,
+              class StructureManagers>
+    inline math::Matrix_t
+    compute_helper(const std::string & representation_name,
+                   const StructureManagers & managersA,
+                   const StructureManagers & managersB) {
+      using internal::KernelType;
+
+      switch (this->kernel_type) {
+      case KernelType::Cosine: {
+        auto kernel = downcast_kernel_impl<KernelType::Cosine>(kernel_impl);
```

Comment:
I am following the same structure as in the calculators here. It is not really needed since the call here is expansive so virtual functions are not an issue. But I felt having too many ways to deal with the same problem in the code would make it even more confusing...
What do you think ?